design/trait-work-as-patch-of-workload.md

```diff
@@ -0,0 +1,256 @@
+# Trait Work as Patch of Workload
+
+- Owner: Shibo Li (@ifree613), Jianbo Sun (@wonderflow)
+- Date: 07/18/2020
+- Status: Draft
+
+## Background
+
+OAM encourages separate of concern for developers and operators. So we use abstract workload like ContainerizedWorkload
+to hide details for K8s built-in object like Deployment. We separate Deployment into one `ContainerizedWorkload` and
+several traits in implementation, both of them have CRDs and controllers.
+
+Everything goes well until we change traits which will affect fields in pod template of Deployment, that will trigger
+re-create of pods. Trait will take effect asynchronously after workload created. In another world, a deployment will be
+created first, after that trait will update the deployment to take effect. Updating pod template of deployment leads to
+re-create of pods, which finally makes the Application unstable. 
+ 
+For example, if we currently have an OAM app running like below:
+
+It just contains a component with ContainerizedWorkload, and a canary trait.
+ 
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: example-component
+spec:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+        - name: wordpress
+          image: wordpress:4.6.1-apache
+          ports:
+            - containerPort: 80
+              name: wordpress
+          env:
+            - name: TEST_ENV
+              value: test
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  - componentName: example-component
+      traits:
+        - trait:
+            apiVersion: extended.oam.dev/v1alpha2
+            kind: Canary
+            spec:
+              canaryNumber: 1
+```
+
+The real running workload is like below:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+spec:
+  template:
+    spec:
+      containers:
+        - name: wordpress
+          image: wordpress:4.6.1-apache
+          ports:
+            - containerPort: 80
+              name: wordpress
+          env:
+            - name: TEST_ENV
+              value: test
+```
+
+After some time, the App operator want to add two traits on this App for some security reason.
```

Comment:
Please clearly define what is "patch trait" in this proposal - this is the most important part.

And please give abundant examples for "patch trait", for example: is ManualScaler a patch trait? is SidecarSet a patch trait? is Service Binding trait a patch trait?